### DISCUSSION TITLE
Unit move schema

### DISCUSSION THREAD

**Discussion by brittonhayes** [brittonhayes/warhammer-aos#23] [General]
As it stands currently, we've got a mixture of data types under a unit's `move` field. For some units we've got and array of objects detailing their movement stages after undergoing damage, for other units we've got a simple string detailing the number of spaces the unit can move.

See https://github.com/brittonhayes/warhammer-aos/issues/22#issue-812980118

This is a breaking issue for unmarshalling the JSON, but I feel like also worthy of chatting about which way feels better. 

How should the JSON be structured for a unit's `move` attribute?

**Comment by Zferg** [comment]
Unless we have all the fields be arrays this will continue to be a problem as, just off the top of my head, there are units with variable move, attack, and save attributes that are affected by the wound track.

I propose that we instead have an attribute under the unit, called something like unit.wound_track, and have a string saying "Detailed to the wound track" in the place of the affected attribute.

**Reply by Zferg** [reply]
One problem would be pulling and parsing that wound track into the API could be a pain as there could be unit-specific attributes in that wound track.

**Reply by brittonhayes** [reply]
I think we could use the D&D 5e SRD API as reference here.

So for sub-items, they use this index object where they indicate the path where you can find the expanded details. This makes a lot of sense for resources that could have properties that are in common for multiple units.

So for elaborated details like this now that an array would add quite a bit of verbosity to each file.

For reference [D&D API](https://www.dnd5eapi.co/api/)
https://www.dnd5eapi.co/api/ability-scores

```
"school": {
		"index": "evocation",
		"name": "Evocation",
		"url": "/api/magic-schools/evocation"
	},
	"classes": [
		{
			"index": "wizard",
			"name": "Wizard",
			"url": "/api/classes/wizard"
		}
	],
	"subclasses": [
		{
			"index": "lore",
			"name": "Lore",
			"url": "/api/subclasses/lore"
		},
		{
			"index": "land",
			"name": "Land",
			"url": "/api/subclasses/land"
		}
	],
	"url": "/api/spells/acid-arrow"
```

**Reply by Zferg** [reply]
Didnt see your reply before sending this before. Each wound track is different for each model in addition to changing names and specifics for said tracks that would only apply to that specific model. For example in the kharadron overlords frigate, the wound track changes the "move" (which is a standard name but the number is not universal across all wound tracks) but also changes the "Boarding Weapons" attack value (whose name is not shared by any other model in the game but one and the other model's boarding weapons don't share the same values).

**Reply by brittonhayes** [reply]
Okay then given that these attributes are completely different per unit, it could be odd to host their details on a different endpoint. I could be expanding the scope of our issue too far here. I think what makes the most sense is an array for each of these multi-value attributes such as move and attack.

**Reply by brittonhayes** [reply]
Here is a unit's pdf for reference 
https://www.games-workshop.com/resources/PDF/AoS_Warscrolls/aos-warscroll-Arkanaut-Frigate-en.pdf

**Reply by Zferg** [reply]
We could just do something like this:
```
"army": "Kharadron Overlords",
  "units": [
    {
      "name": "Aether-Khemist",
      "size": "32 mm",
      "move": "see damage track",
      "save": "see damage track",
      "bravery": "7",
      "wounds": "5",
      "missile_weapon": [
        { 
          "name": "Atmospheric Anatomiser",
          "range": "9",
          "attacks": "see damage track",
          "to_hit": "4+",
          "to_wound": "4+",
          "rend": "-2",
          "damage": "1"
         }
      ],
      "melee_weapon": [
        { }
      ],
      "abilites": [
        { }
      ],
      "damage_track" : [
        [ INSERT DAMAGE TRACK HERE ],
        "
      "keywords": [
        { }
      ]
    },
```

**Reply by Zferg** [reply]
and then you would only need to add an "if damage_track is present then pull" to the api

**Reply by brittonhayes** [reply]
Okay, I like this idea. Because that keeps the top level details very simple but then allows the fields to be of a consistent data type. String for `move` and `[]DamageTrack` for the damage track. I think this is worth trying out on a file.
